auth/auth.go

```diff
@@ -0,0 +1,272 @@
+// subsequent calls to this function do not perform an auth dance every single
+// time.
+func (h *GoogleHeadersProvider) getGoogleAccessToken(ctx context.Context) (*oauth2.Token, error) {
+	if h.token != nil && h.token.Valid() {
+		return h.token, nil
+	}
+
+	serviceAccountJSON, err := convertYAMLStructToEncodedJSON(&h.authOptions)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	credentials, err := google.CredentialsFromJSON(ctx, serviceAccountJSON, "email", "profile", "openid", groupScope)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	h.token, err = credentials.TokenSource.Token()
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	return h.token, nil
+}
+
+// GetHeaders returns a map overwriting the authorization header.
+func (h *MacaroonHeadersProvider) GetHeaders(ctx context.Context) (map[string]string, error) {
+	ms, err := h.getDischargedMacaroon(ctx)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	buf, err := ms.MarshalBinary()
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	headerMacaroon := base64.RawURLEncoding.EncodeToString(buf)
+	return map[string]string{"authorization": "Macaroon " + headerMacaroon}, nil
+}
+
+// getDischargedMacaroon returns a discharged macaroon to be used for
+// authorization with the Temporal Server. It fetches a macaroon from the Server
+// and discharges it against the auth provider specified in it.
+//
+// It also caches the macaroon in memory, for as long as it is valid, such that
+// subsequent calls to this function do not perform an auth dance every single
+// time.
+func (h *MacaroonHeadersProvider) getDischargedMacaroon(ctx context.Context) (macaroon.Slice, error) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.ms != nil {
+		expiry, found := checkers.MacaroonsExpiryTime(nil, h.ms)
+		if !found || time.Now().Before(expiry) {
+			return h.ms, nil
+		}
+	}
+
+	resp, err := h.bakeryClient.Get(h.authOptions.MacaroonURL)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	json, err := base64.RawURLEncoding.DecodeString(string(body))
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	var m bakery.Macaroon
+	if err = m.UnmarshalJSON(json); err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	caveats := m.M().Caveats()
+	if len(caveats) == 0 {
+		return nil, errgo.Newf("macaroon missing required caveat")
+	}
+	authURL := caveats[0].Location
+
+	if err := agent.SetUpAuth(h.bakeryClient, &agent.AuthInfo{
+		Key: h.authOptions.AgentKey,
+		Agents: []agent.Agent{{
+			URL:      authURL,
+			Username: h.authOptions.AgentUsername,
+		}},
+	}); err != nil {
+		return nil, errgo.Mask(err)
+	}
+	h.ms, err = h.bakeryClient.DischargeAll(ctx, &m)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+
+	return h.ms, nil
+}
+
+func NewAuthHeadersProvider(opts *AuthOptions) (HeadersProvider, error) {
```

Comment:
General question: should we add godocs for all the exposed functions & methods in the Temporal lib? I would rather we stick to one approach and used it in this repository, but there is also the "C3: Redundant Comment" code smell, so please use your judgement.